File: bindings/go/model_manager_test.go
package blazen

import (
	"context"
	"errors"
	"testing"
	"time"
)

func newTestManagerCtx(t *testing.T) (*ModelManager, context.Context, context.CancelFunc) {
	t.Helper()
	mm := NewModelManager()
	t.Cleanup(mm.Close)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	return mm, ctx, cancel
}

func TestModelManagerConstructors(t *testing.T) {
	t.Run("default", func(t *testing.T) {
		mm := NewModelManager()
		if mm == nil {
			t.Fatal("NewModelManager returned nil")
		}
		mm.Close()
	})

	t.Run("budgets_gb", func(t *testing.T) {
		mm := NewModelManagerWithBudgetsGB(8, 16)
		if mm == nil {
			t.Fatal("NewModelManagerWithBudgetsGB returned nil")
		}
		mm.Close()
	})

	t.Run("pool_budgets", func(t *testing.T) {
		mm, err := NewModelManagerWithPoolBudgets(map[string]float64{
			"cpu":   4,
			"gpu:0": 12,
		})
		if err != nil {
			t.Fatalf("NewModelManagerWithPoolBudgets: %v", err)
		}
		if mm == nil {
			t.Fatal("NewModelManagerWithPoolBudgets returned nil")
		}
		mm.Close()
	})

	t.Run("pool_budgets_invalid_label", func(t *testing.T) {
		_, err := NewModelManagerWithPoolBudgets(map[string]float64{
			"not-a-real-pool": 1,
		})
		if err == nil {
			t.Fatal("expected error for invalid pool label")
		}
	})
}

func TestModelManagerStatusEmpty(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	status, err := mm.Status(ctx)
	if err != nil {
		t.Fatalf("Status: %v", err)
	}
	if len(status) != 0 {
		t.Fatalf("expected empty status, got %d entries: %+v", len(status), status)
	}
}

func TestModelManagerPoolsDefault(t *testing.T) {
	mm := NewModelManager()
	t.Cleanup(mm.Close)

	pools := mm.Pools()
	if len(pools) != 2 {
		t.Fatalf("expected 2 default pools (cpu + gpu:0), got %d: %+v", len(pools), pools)
	}
	seen := map[string]bool{}
	for _, p := range pools {
		seen[p.Pool] = true
		if p.BudgetBytes == 0 {
			t.Errorf("pool %q has zero budget; default constructor should seed u64::MAX", p.Pool)
		}
	}
	if !seen["cpu"] {
		t.Errorf("default pools missing %q: %+v", "cpu", pools)
	}
	if !seen["gpu:0"] {
		t.Errorf("default pools missing %q: %+v", "gpu:0", pools)
	}
}

func TestModelManagerLoadNonexistent(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	err := mm.Load(ctx, "definitely-not-registered")
	if err == nil {
		t.Fatal("expected error loading unregistered model")
	}
	var be Error
	if !errors.As(err, &be) {
		t.Fatalf("expected typed blazen.Error, got %T: %v", err, err)
	}
}

func TestModelManagerIsLoadedNonexistent(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	loaded, err := mm.IsLoaded(ctx, "definitely-not-registered")
	if err != nil {
		t.Fatalf("IsLoaded: %v", err)
	}
	if loaded {
		t.Fatal("IsLoaded for unregistered model should be false")
	}
}

func TestModelManagerLoadAdapterNonexistent(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	_, err := mm.LoadAdapter(ctx, "definitely-not-registered", "/nonexistent/adapter/dir", AdapterOptions{Scale: 1.0})
	if err == nil {
		t.Fatal("expected error loading adapter for unregistered model")
	}
	var be Error
	if !errors.As(err, &be) {
		t.Fatalf("expected typed blazen.Error, got %T: %v", err, err)
	}
}

func TestModelManagerListAdaptersNonexistent(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	_, err := mm.ListAdapters(ctx, "definitely-not-registered")
	if err == nil {
		t.Fatal("expected error listing adapters for unregistered model")
	}
}

func TestModelManagerPoolBytes(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	used, err := mm.UsedBytes(ctx, "cpu")
	if err != nil {
		t.Fatalf("UsedBytes(cpu): %v", err)
	}
	if used != 0 {
		t.Errorf("expected UsedBytes(cpu)=0 on empty manager, got %d", used)
	}

	if _, err := mm.AvailableBytes(ctx, "cpu"); err != nil {
		t.Fatalf("AvailableBytes(cpu): %v", err)
	}

	if _, err := mm.UsedBytes(ctx, "not-a-pool"); err == nil {
		t.Fatal("expected error from UsedBytes with invalid pool label")
	}
}

func TestModelManager_LoadFromHf_NoNetwork(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	// Why: an obviously-invalid repo slug must fail with a typed Blazen
	// error rather than panic. Either the loader rejects validation
	// up-front or the HF probe surfaces a network/404 error — both
	// outcomes are acceptable as long as the call returns cleanly.
	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("LoadFromHf panicked on invalid repo: %v", r)
		}
	}()

	backend, err := mm.LoadFromHf(ctx, "test-id", "nonexistent/repo-abc123-xyz", HfLoadOptions{})
	if err == nil {
		t.Fatalf("expected error loading nonexistent repo, got backend=%q", backend)
	}
	var be Error
	if !errors.As(err, &be) {
		t.Fatalf("expected typed blazen.Error, got %T: %v", err, err)
	}
}

func TestModelManager_LoadFromHf_NilOptions_DoesntPanic(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("LoadFromHf panicked on zero-value HfLoadOptions: %v", r)
		}
	}()

	_, err := mm.LoadFromHf(ctx, "test-id-empty-opts", "nonexistent/repo-abc123-xyz", HfLoadOptions{})
	// Why: success would be unexpected for a fake repo, but we accept it
	// — the contract is "no panic". Any error path is also fine.
	_ = err
}

func TestModelManager_LoadFromHf_AllOptionsPlumbed(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	hint := BackendHintLlamacpp
	rev := "main"
	token := "hf_dummy"
	cache := "/tmp/blazen-hf-cache-test"
	device := "cpu"
	gguf := "model-q4.gguf"
	var mem uint64 = 1 << 30
	pool := "cpu"
	opts := HfLoadOptions{
		BackendHint:         &hint,
		Revision:            &rev,
		HfToken:             &token,
		CacheDir:            &cache,
		Device:              &device,
		GgufFile:            &gguf,
		MemoryEstimateBytes: &mem,
		Pool:                &pool,
	}

	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("LoadFromHf panicked with all options set: %v", r)
		}
	}()
	if _, err := mm.LoadFromHf(ctx, "test-id-all-opts", "nonexistent/repo-abc123-xyz", opts); err == nil {
		t.Fatal("expected error loading nonexistent repo")
	}
}

func TestModelManager_LoadFromHf_AfterClose(t *testing.T) {
	mm := NewModelManager()
	mm.Close()

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	_, err := mm.LoadFromHf(ctx, "id", "owner/repo", HfLoadOptions{})
	if err == nil {
		t.Fatal("expected error from LoadFromHf after Close")
	}
	var ve *ValidationError
	if !errors.As(err, &ve) {
		t.Fatalf("expected *ValidationError after Close, got %T: %v", err, err)
	}
}

func newTrainConfig(maxSteps uint32) TrainConfig {
	return TrainConfig{
		BaseModelRepo: "nonexistent/repo-abc123-xyz",
		OutputDir:     "/tmp/blazen-train-out-test",
		Lora: LoraConfig{
			Rank:          8,
			Alpha:         16,
			Dropout:       0.05,
			TargetModules: []string{"q_proj", "v_proj"},
		},
		Optim: OptimConfig{
			LearningRate: 2e-4,
			Beta1:        0.9,
			Beta2:        0.999,
			Epsilon:      1e-8,
			WeightDecay:  0.0,
		},
		Scheduler: SchedulerConfig{
			Kind:        SchedulerKindCosine,
			WarmupSteps: 0,
		},
		MaxSteps:                  maxSteps,
		BatchSize:                 1,
		GradientAccumulationSteps: 1,
		MaxSeqLen:                 64,
		Seed:                      42,
		MixedPrecision:            MixedPrecisionNone,
	}
}

func TestTrainLora_RejectsInvalidConfig(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	ds, err := NewJsonlDatasetFromPath("/nonexistent/data.jsonl", "/nonexistent/tok.json", nil, 64, nil, 0)
	// Why: the dataset constructor must fail because the path does not
	// exist; we still cover the TrainLora invalid-config path below by
	// passing a nil dataset, which TrainLora must reject up-front.
	if err == nil {
		t.Fatal("expected NewJsonlDatasetFromPath to error on bogus path")
	}
	if ds != nil {
		t.Fatalf("expected nil dataset on error, got %+v", ds)
	}

	_, err = mm.TrainLora(ctx, newTrainConfig(0), nil, nil)
	if err == nil {
		t.Fatal("expected TrainLora to reject MaxSteps=0 / nil dataset")
	}
	var be Error
	if !errors.As(err, &be) {
		t.Fatalf("expected typed blazen.Error, got %T: %v", err, err)
	}
}

func TestTrainLora_NilProgress_DoesntPanic(t *testing.T) {
	mm, ctx, cancel := newTestManagerCtx(t)
	defer cancel()

	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("TrainLora panicked with nil progress: %v", r)
		}
	}()

	// Why: we have no real dataset to construct, so we expect the call
	// to fail at the dataset-validation guard rather than launch a
	// training run. The contract under test is "no panic".
	_, err := mm.TrainLora(ctx, newTrainConfig(1), nil, nil)
	if err == nil {
		t.Fatal("expected error from TrainLora with nil dataset")
	}
}

func TestJsonlDataset_RequiresValidPath(t *testing.T) {
	_, err := NewJsonlDatasetFromPath("/nonexistent/file.jsonl", "/nonexistent/tok.json", nil, 64, nil, 0)
	if err == nil {
		t.Fatal("expected NewJsonlDatasetFromPath to error on missing file")
	}
	var be Error
	if !errors.As(err, &be) {
		t.Fatalf("expected typed blazen.Error, got %T: %v", err, err)
	}
}

func TestModelManagerCloseIsIdempotent(t *testing.T) {
	mm := NewModelManager()
	mm.Close()
	mm.Close()
	mm.Close()
}

func TestModelManagerOperationsAfterClose(t *testing.T) {
	mm := NewModelManager()
	mm.Close()

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	if err := mm.Load(ctx, "anything"); err == nil {
		t.Error("expected error from Load after Close")
	}
	if _, err := mm.IsLoaded(ctx, "anything"); err == nil {
		t.Error("expected error from IsLoaded after Close")
	}
	if _, err := mm.Status(ctx); err == nil {
		t.Error("expected error from Status after Close")
	}
	if pools := mm.Pools(); pools != nil {
		t.Errorf("expected nil Pools after Close, got %+v", pools)
	}
}

File: bindings/kotlin/src/test/kotlin/dev/zorpx/blazen/ModelManagerTest.kt
package dev.zorpx.blazen

import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import dev.zorpx.blazen.uniffi.BlazenException as UniffiBlazenException

/**
 * Smoke tests for the idiomatic [ModelManager] wrapper.
 *
 * Mirrors the Go binding's `TestModelManager*` shape: exercise the three
 * constructors, snapshot verbs against an empty manager, and assert that
 * verbs operating on unknown ids raise the generated
 * [UniffiBlazenException] sealed exception so callers can `when`-match on
 * the concrete variant.
 */
class ModelManagerTest {
    @Test
    fun `default constructor seeds cpu and gpu0 pools with no budget enforcement`() {
        ModelManager().use { mgr ->
            val pools = mgr.pools().associateBy { it.pool }
            assertNotNull(pools["cpu"], "cpu pool must be present, got: $pools")
            assertNotNull(pools["gpu:0"], "gpu:0 pool must be present, got: $pools")
            assertEquals(ULong.MAX_VALUE, pools.getValue("cpu").budgetBytes)
            assertEquals(ULong.MAX_VALUE, pools.getValue("gpu:0").budgetBytes)
        }
    }

    @Test
    fun `gigabyte budget constructor records both pool budgets`() {
        ModelManager(cpuRamGb = 4.0, gpuVramGb = 8.0).use { mgr ->
            val pools = mgr.pools().associateBy { it.pool }
            // Why: 1 GiB == 1_073_741_824 bytes, matching the Rust constructor's
            // power-of-two conversion (not 1e9). Asserting the exact byte count
            // pins the convention the wrapper inherits from the upstream crate.
            assertEquals(4uL * 1_073_741_824uL, pools.getValue("cpu").budgetBytes)
            assertEquals(8uL * 1_073_741_824uL, pools.getValue("gpu:0").budgetBytes)
        }
    }

    @Test
    fun `per-pool budget constructor accepts custom labels`() {
        ModelManager(poolBudgets = mapOf("cpu" to 2.0, "gpu:1" to 6.0)).use { mgr ->
            val pools = mgr.pools().associateBy { it.pool }
            assertEquals(2uL * 1_073_741_824uL, pools.getValue("cpu").budgetBytes)
            assertEquals(6uL * 1_073_741_824uL, pools.getValue("gpu:1").budgetBytes)
        }
    }

    @Test
    fun `status returns empty list on a fresh manager`() = runBlocking {
        ModelManager().use { mgr ->
            assertEquals(emptyList<ModelStatus>(), mgr.status())
        }
    }

    @Test
    fun `isLoaded returns false for unknown model id`() = runBlocking {
        ModelManager().use { mgr ->
            assertEquals(false, mgr.isLoaded("does-not-exist"))
        }
    }

    @Test
    fun `load on an unregistered model id raises a Blazen exception`() {
        ModelManager().use { mgr ->
            val ex = assertThrows(UniffiBlazenException::class.java) {
                runBlocking { mgr.load("does-not-exist") }
            }
            assertNotNull(ex.message)
            assertTrue(ex.message!!.isNotEmpty(), "exception message must be non-empty")
        }
    }

    @Test
    fun `loadAdapter on an unregistered model id raises a Blazen exception`() {
        ModelManager().use { mgr ->
            val ex = assertThrows(UniffiBlazenException::class.java) {
                runBlocking {
                    mgr.loadAdapter(
                        modelId = "does-not-exist",
                        adapterDir = "/nonexistent/lora",
                        options = AdapterOptions(adapterId = "test-adapter"),
                    )
                }
            }
            assertNotNull(ex.message)
        }
    }

    @Test
    fun `AdapterOptions default adapter id is a non-empty unique string`() {
        val a = AdapterOptions()
        val b = AdapterOptions()
        assertTrue(a.adapterId.isNotEmpty())
        assertTrue(b.adapterId.isNotEmpty())
        // Why: defaults are UUIDs — two adjacent constructions must not
        // collide or the contract that callers can mount adapters without
        // inventing names breaks.
        assertTrue(a.adapterId != b.adapterId, "default adapter ids must be unique")
        assertEquals(1.0, a.scale)
    }

    @Test
    fun `loadFromHf throws on nonexistent repo`() {
        ModelManager().use { mgr ->
            val ex = assertThrows(UniffiBlazenException::class.java) {
                runBlocking {
                    mgr.loadFromHf(
                        id = "ghost",
                        repo = "this-org-does-not-exist/this-repo-also-does-not-exist",
                    )
                }
            }
            assertNotNull(ex.message)
            assertTrue(ex.message!!.isNotEmpty(), "exception message must be non-empty")
        }
    }

    @Test
    fun `loadFromHf with default options does not crash`() {
        // Why: a default-constructed HfLoadOptions must round-trip through
        // the FFI without panicking on the all-null record. We expect a
        // BlazenException (empty repo id), but never a JVM crash or a
        // NullPointerException from the converter — that is the bug we
        // are guarding against.
        ModelManager().use { mgr ->
            val ex = assertThrows(UniffiBlazenException::class.java) {
                runBlocking {
                    mgr.loadFromHf(id = "empty", repo = "")
                }
            }
            assertNotNull(ex.message)
        }
    }

    @Test
    fun `BackendHint values map to stable lower-case wire strings`() {
        assertEquals("mistralrs", BackendHint.MISTRALRS.value)
        assertEquals("candle", BackendHint.CANDLE.value)
        assertEquals("llamacpp", BackendHint.LLAMACPP.value)
    }

    @Test
    fun `trainLora rejects invalid config with maxSteps of zero`() {
        ModelManager().use { mgr ->
            val tokenizer = java.io.File.createTempFile("blazen-train-kotlin-wrap-tok", ".json").apply {
                writeText("{\"version\":\"1.0\",\"model\":{\"type\":\"BPE\",\"vocab\":{},\"merges\":[]}}")
                deleteOnExit()
            }
            val data = java.io.File.createTempFile("blazen-train-kotlin-wrap-data", ".jsonl").apply {
                writeText("{\"text\":\"hi\"}\n")
                deleteOnExit()
            }
            val outDir = java.nio.file.Files.createTempDirectory("blazen-train-kotlin-wrap-out").toFile().apply {
                deleteOnExit()
            }
            // Why: dataset construction may itself fail before train_lora is reached if
            // the tokenizer is too minimal — both failure modes prove the FFI surface
            // refuses the invalid configuration rather than crashing the JVM.
            val ex = assertThrows(UniffiBlazenException::class.java) {
                runBlocking {
                    val ds = JsonlDataset.fromPath(
                        path = data.absolutePath,
                        tokenizerPath = tokenizer.absolutePath,
                    )
                    mgr.trainLora(
                        config = TrainConfig(
                            baseModelRepo = "hf-internal-testing/tiny-random-gpt2",
                            outputDir = outDir.absolutePath,
                            maxSteps = 0u,
                        ),
                        dataset = ds,
                    )
                }
            }
            assertNotNull(ex.message)
        }
    }

    @Test
    fun `JsonlDataset throws on invalid path`() {
        val ex = assertThrows(UniffiBlazenException::class.java) {
            JsonlDataset.fromPath(
                path = "/nonexistent/blazen-train-kotlin-wrap/missing.jsonl",
                tokenizerPath = "/nonexistent/blazen-train-kotlin-wrap/missing-tokenizer.json",
            )
        }
        assertNotNull(ex.message)
        assertTrue(ex.message!!.isNotEmpty(), "exception message must be non-empty")
    }

    @Test
    fun `TrainingEvent exhaustive when compiles`() {
        // Why: this test exists solely to lock the sealed-class shape in
        // place. If a new variant is added upstream without updating the
        // wrapper's TrainingEvent and conversion, the compiler will flag
        // a missing branch here and fail the test build.
        val events: List<TrainingEvent> = listOf(
            TrainingEvent.Started(totalSteps = 10uL),
            TrainingEvent.StepCompleted(step = 1uL, loss = 0.5f, learningRate = 2e-4, elapsedMs = 100uL),
            TrainingEvent.Evaluating(step = 5uL),
            TrainingEvent.EvalCompleted(step = 5uL, evalLoss = 0.4f),
            TrainingEvent.CheckpointSaved(step = 5uL, path = "/tmp/ckpt"),
            TrainingEvent.Finished(finalLoss = 0.3f, totalSteps = 10uL, adapterDir = "/tmp/adapter"),
        )
        val labels = events.map { ev ->
            when (ev) {
                is TrainingEvent.Started -> "started:${ev.totalSteps}"
                is TrainingEvent.StepCompleted -> "step:${ev.step}"
                is TrainingEvent.Evaluating -> "evaluating:${ev.step}"
                is TrainingEvent.EvalCompleted -> "eval:${ev.step}"
                is TrainingEvent.CheckpointSaved -> "ckpt:${ev.step}"
                is TrainingEvent.Finished -> "finished:${ev.totalSteps}"
            }
        }
        assertEquals(6, labels.size)
        assertEquals("started:10", labels[0])
        assertEquals("finished:10", labels[5])
    }

    @Test
    fun `TrainConfig defaults match documented surface`() {
        val cfg = TrainConfig(baseModelRepo = "repo", outputDir = "/tmp/out")
        assertEquals(1000u, cfg.maxSteps)
        assertEquals(4u, cfg.batchSize)
        assertEquals(2048u, cfg.maxSeqLen)
        assertEquals(42uL, cfg.seed)
        assertEquals(MixedPrecision.BF16, cfg.mixedPrecision)
        assertEquals(SchedulerKind.COSINE, cfg.scheduler.kind)
        assertEquals(16u, cfg.lora.rank)
        assertEquals(32.0f, cfg.lora.alpha)
        assertEquals(listOf("q_proj", "k_proj", "v_proj", "o_proj"), cfg.lora.targetModules)
        assertEquals(2e-4, cfg.optim.learningRate)
        assertEquals(1.0f, cfg.optim.gradientClip)
    }

    @Test
    fun `SchedulerKind and MixedPrecision wire strings are stable`() {
        assertEquals("constant", SchedulerKind.CONSTANT.value)
        assertEquals("linear", SchedulerKind.LINEAR.value)
        assertEquals("cosine", SchedulerKind.COSINE.value)
        assertEquals("none", MixedPrecision.NONE.value)
        assertEquals("bf16", MixedPrecision.BF16.value)
    }

    @Test
    fun `DpoConfig defaults match documented surface`() {
        val cfg = DpoConfig(core = TrainCoreConfig(baseModelRepo = "repo", outputDir = "/tmp/out"))
        assertEquals(0.1f, cfg.beta)
        assertEquals(0.0f, cfg.labelSmoothing)
        assertEquals(null, cfg.referenceModelRepo)
        assertEquals(null, cfg.referenceModelRevision)
        // Core defaults: smaller per-step batches than TrainConfig because
        // each preference step carries two forward passes (chosen + rejected).
        assertEquals(1u, cfg.core.batchSize)
        assertEquals(8u, cfg.core.gradientAccumulationSteps)
        assertEquals(1024u, cfg.core.maxSeqLen)
        assertEquals(MixedPrecision.BF16, cfg.core.mixedPrecision)
        // LoRA defaults reused from PR7's LoraConfig.
        assertEquals(16u, cfg.lora.rank)
        assertEquals(32.0f, cfg.lora.alpha)
    }

    @Test
    fun `OrpoConfig defaults match documented surface`() {
        val cfg = OrpoConfig(core = TrainCoreConfig(baseModelRepo = "repo", outputDir = "/tmp/out"))
        assertEquals(0.1f, cfg.lambda)
        assertEquals(16u, cfg.lora.rank)
        assertEquals(MixedPrecision.BF16, cfg.core.mixedPrecision)
    }

    @Test
    fun `SimpoConfig defaults follow TRL main`() {
        val cfg = SimpoConfig(core = TrainCoreConfig(baseModelRepo = "repo", outputDir = "/tmp/out"))
        // Why: TRL `main` (and the upstream Rust default) sets beta=2.0,
        // gamma=1.0 — pinning these guards against silent drift if the
        // generated record changes its defaults.
        assertEquals(2.0f, cfg.beta)
        assertEquals(1.0f, cfg.gamma)
    }

    @Test
    fun `KtoConfig defaults weight both signs equally`() {
        val cfg = KtoConfig(core = TrainCoreConfig(baseModelRepo = "repo", outputDir = "/tmp/out"))
        assertEquals(0.1f, cfg.beta)
        // Why: lambda_D and lambda_U both default to 1.0, so the desirable
        // and undesirable terms contribute equally before per-sign tuning.
        assertEquals(1.0f, cfg.lambdaD)
        assertEquals(1.0f, cfg.lambdaU)
        assertEquals(null, cfg.referenceModelRepo)
    }

    @Test
    fun `FullFineTuneConfig defaults expose gradient checkpointing flag`() {
        val cfg = FullFineTuneConfig(core = TrainCoreConfig(baseModelRepo = "repo", outputDir = "/tmp/out"))
        // Why: gradient_checkpointing is accepted for forward compatibility
        // but the Rust trainer rejects `true` at init time on candle 0.10.2.
        // Default must be `false` so callers don't accidentally trip that
        // validation.
        assertEquals(false, cfg.gradientCheckpointing)
        assertEquals(1u, cfg.core.batchSize)
    }

    @Test
    fun `PreferenceJsonlDataset throws on invalid path`() {
        val ex = assertThrows(UniffiBlazenException::class.java) {
            PreferenceJsonlDataset.fromPath(
                path = "/nonexistent/blazen-train-kotlin-wrap/missing-pref.jsonl",
                tokenizerPath = "/nonexistent/blazen-train-kotlin-wrap/missing-tokenizer.json",
            )
        }
        assertNotNull(ex.message)
        assertTrue(ex.message!!.isNotEmpty(), "exception message must be non-empty")
    }

    @Test
    fun `RatedJsonlDataset throws on invalid path`() {
        val ex = assertThrows(UniffiBlazenException::class.java) {
            RatedJsonlDataset.fromPath(
                path = "/nonexistent/blazen-train-kotlin-wrap/missing-rated.jsonl",
                tokenizerPath = "/nonexistent/blazen-train-kotlin-wrap/missing-tokenizer.json",
            )
        }
        assertNotNull(ex.message)
        assertTrue(ex.message!!.isNotEmpty(), "exception message must be non-empty")
    }

    @Suppress("UNUSED_VARIABLE", "unused")
    @Test
    fun `trainDpo trainOrpo trainSimpo trainKto fineTune signatures compile`() {
        // Why: pure compile-time guard. We never invoke these references —
        // the suspending native call would block on download/init — but
        // assigning them to typed function values forces the compiler to
        // resolve each overload, so any drift in the wrapper signature
        // (parameter names, optional progress callback, return type)
        // surfaces as a build failure rather than a runtime surprise.
        val mgr: ModelManager? = null
        if (mgr != null) {
            val dpo: suspend (DpoConfig, PreferenceJsonlDataset, ((TrainingEvent) -> Unit)?) -> TrainedAdapter =
                mgr::trainDpo
            val orpo: suspend (OrpoConfig, PreferenceJsonlDataset, ((TrainingEvent) -> Unit)?) -> TrainedAdapter =
                mgr::trainOrpo
            val simpo: suspend (SimpoConfig, PreferenceJsonlDataset, ((TrainingEvent) -> Unit)?) -> TrainedAdapter =
                mgr::trainSimpo
            val kto: suspend (KtoConfig, RatedJsonlDataset, ((TrainingEvent) -> Unit)?) -> TrainedAdapter =
                mgr::trainKto
            val ft: suspend (FullFineTuneConfig, JsonlDataset, ((TrainingEvent) -> Unit)?) -> FullFineTuneResult =
                mgr::fineTune
        }
        // Pure compile-time guard — no runtime assertion needed.
        assertTrue(true)
    }

    @Test
    fun `ForeignLocalModel interface is implementable from Kotlin`() {
        // Why: the interface must be openly implementable (not sealed, not
        // requiring internal types) — exercise that by constructing an
        // anonymous implementation here. The instance is never registered;
        // we are only asserting the interface shape compiles.
        val model = object : ForeignLocalModel {
            override suspend fun load() = Unit
            override suspend fun unload() = Unit
            override suspend fun isLoaded(): Boolean = false
            override fun device(): String = "cpu"
            override suspend fun memoryBytes(): ULong? = null
            override suspend fun loadAdapter(
                adapterDir: String,
                options: AdapterOptions,
            ): AdapterHandle = AdapterHandle(options.adapterId, 0uL, "attached")
            override suspend fun unloadAdapter(handle: AdapterHandle) = Unit
            override suspend fun listAdapters(): List<AdapterStatus> = emptyList()
        }
        assertEquals("cpu", model.device())
    }
}
